### ISSUE TITLE
Ramp_fit error on MIRI NGROUPS<=2 exposure

### ISSUE THREAD

**Issue by hbushouse** [STScI-JWST/jwst#2035]
DMS testing of 0.9.3rc1 reports:
```
/ifs/int/jwstc/store/doggett/tests/QnGV2/cstring/JWSTDP-2018_1-180430/2018-05-16-145153/sdp/fits/jw93135336001_02102_00001_mirifulong_uncal.fits
 
CRDS - DEBUG - Final effective context is 'jwst_0456.pmap' 
CRDS - DEBUG - Computing best references locally. 
CRDS - DEBUG - Bestrefs header: 
{'META.EXPOSURE.READPATT [READPATT]': 'FAST', 
'META.EXPOSURE.TYPE [EXP_TYPE]': 'MIR_FLATMRS', 
'META.INSTRUMENT.BAND [BAND]': 'LONG', 
'META.INSTRUMENT.CHANNEL [CHANNEL]': '34.0', 
'META.INSTRUMENT.DETECTOR [DETECTOR]': 'MIRIFULONG', 
'META.INSTRUMENT.FILTER [FILTER]': 'N/A', 
'META.INSTRUMENT.NAME [INSTRUME]': 'MIRI', 
'META.OBSERVATION.DATE [DATE-OBS]': '2016-01-23', 
'META.OBSERVATION.TIME [TIME-OBS]': '13:55:06.059', 
'META.SUBARRAY.NAME [SUBARRAY]': 'FULL', 
'REFTYPE': 'UNDEFINED'} 
CRDS - DEBUG - Reference type 'gain' defined as 'jwst_miri_gain_0006.fits' 
2018-05-16 15:30:39,186 - stpipe.Detector1Pipeline.ramp_fit - INFO - Using READNOISE reference file: /ifs/int/jwstc/ref/crds/cache/references/jwst/miri/jwst_miri_readnoise_0038.fits 
2018-05-16 15:30:39,403 - stpipe.Detector1Pipeline.ramp_fit - INFO - Using GAIN reference file: /ifs/int/jwstc/ref/crds/cache/references/jwst/miri/jwst_miri_gain_0006.fits 
2018-05-16 15:30:39,482 - stpipe.Detector1Pipeline.ramp_fit - INFO - Using algorithm = ols 
2018-05-16 15:30:39,482 - stpipe.Detector1Pipeline.ramp_fit - INFO - Using weighting = optimal 
2018-05-16 15:30:39,489 - stpipe.Detector1Pipeline.ramp_fit - INFO - MIRI dataset has all final groups flagged as DO_NOT_USE. 
2018-05-16 15:30:39,489 - stpipe.Detector1Pipeline.ramp_fit - ERROR - MIRI datasets require at least 2 groups/integration 
2018-05-16 15:30:39,489 - stpipe.Detector1Pipeline.ramp_fit - ERROR - (NGROUPS), so will not process this dataset. 
Traceback (most recent call last): 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/EGG-INFO/scripts/strun", line 14, in <module> 
step = Step.from_cmdline(sys.argv[1:]) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/stpipe/step.py", line 157, in from_cmdline 
return cmdline.step_from_cmdline(args) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/stpipe/cmdline.py", line 297, in step_from_cmdline 
step.run(*positional) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/stpipe/step.py", line 391, in run 
step_result = self.process(*args) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/pipeline/calwebb_detector1.py", line 132, in process 
input, ints_model = self.ramp_fit(input) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/stpipe/step.py", line 391, in run 
step_result = self.process(*args) 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/ramp_fitting/ramp_fit_step.py", line 69, in process 
self.weighting 
File "/ifs/int/jwstc/pipeline/pkgs/miniconda3/envs/jwstdp-0.9.3rc1-py35/lib/python3.5/site-packages/jwst-0.9.3rc1-py3.5-linux-x86_64.egg/jwst/ramp_fitting/ramp_fit.py", line 101, in ramp_fit 
new_model.meta.bunit_data = 'DN/s' 
AttributeError: 'NoneType' object has no attribute 'meta' 
```
The ramp_fit step is returning without computing any slopes because it doesn't have enough groups per integration to work with. It then fails in the calling routine when trying to set a value for the BUNIT keyword in the output model and apparently the output model is either None or undefined.

Shouldn't the step still be calculating pseudo-slope values for these cases, by just subtracting group 1 from group 2, or group 0 from group 1, and return a populated data model?

**Comment by hbushouse** [comment]
Also, I don't think the messages from ramp_fit about the lack of sufficient groups should be in the ERROR category. These should just be warnings.

**Comment by hbushouse** [comment]
You're right @dmggh - this is not an issue with the ramp_fit step. This needs to be handled in the firstframe and lastframe steps. The exposure in question has NGROUPS=2 and both firstframe and lastframe are only checking to see if NGROUPS>1, so they're both being applied, leaving no good groups to work with in ramp_fit at all. We need to update the firstframe and lastframe steps to comply with the guidelines in #1681.

**Comment by hbushouse** [comment]
What's interesting, though, is that if I run this exposure through `calwebb_detector1` with the `lastframe` step set to skip, but still applying `firstframe`, the `ramp_fit` doesn't seem to complain at all about only having 1 good group per integration to work with and produces some kind of slope result. Theoretically this situation should never be encountered in normal processing (once we update the `firstframe` and `lastframe` steps to be skipped), but you never know what a user might try doing on their own.

**Comment by hbushouse** [comment]
The desired logic for updating the NGROUPS checking in the firstframe and lastframe steps has been documented and confirmed in #1681. In a nutshell:

firstframe: if NGROUPS < 4, skip
lastframe: if NGROUPS < 3, skip

Which means both steps will be skipped for NGROUPS=2, which should fix the return error with ramp_fit, because it will have good data to work with.

**Comment by hbushouse** [comment]
Fixed by #2040.
